### ISSUE TITLE
linter: typescript/explicit-function-return-type false positive on return of function returning a function

### ISSUE THREAD

**Issue by Ragnar-Oock** [oxc-project/oxc#10568]
### What version of Oxlint are you using?

0.16.7

### What command did you run?

oxlint

### What does your `.oxlintrc.json` config file look like?
<details open><summary>`.oxlintrc.json` file content</summary>
<p>

```jsonc
{
  "$schema": "./node_modules/oxlint/configuration_schema.json",
  "categories": {
    "correctness": "error",
    "perf": "warn",
    "suspicious": "warn"
  },
  "plugins": ["unicorn", "typescript", "oxc", "eslint", "import", "promise", "jsdoc"],
  "rules": {
    "no-bitwise": "error",
    "no-div-regex": "error",
    "no-empty": "warn",
    "no-empty-function": "warn",
    "no-eq-null": "warn",
    "no-proto": "error",
    "no-regex-spaces": "error",
    "no-unused-expressions": "warn",
    "no-var": "error",
    "unicode-bom": "error",
    "no-amd": "error",
    "no-commonjs": "error",
    "no-cycle": "error",
    "no-webpack-loader-syntax": "error",
    "unambiguous": "warn",
    "empty-tags": "warn",
    "bad-bitwise-operator": "error",
    "no-const-enum": "error",
    "catch-or-return": "error",
    "spec-only": "error",
    "explicit-function-return-type": "warn",
    "no-dynamic-delete": "warn",
    "no-empty-object-type": "warn",
    "no-explicit-any": "warn",
    "no-import-type-side-effects": "error",
    "no-namespace": "warn",
    "no-non-null-asserted-nullish-coalescing": "warn",
    "no-non-null-assertion": "error",
    "prefer-literal-enum-member": "error",
    "no-abusive-eslint-disable": "warn",
    "no-anonymous-default-export": "error",
    "no-document-cookie": "error",
    "prefer-modern-math-apis": "warn",
    "prefer-node-protocol": "error",
    "prefer-number-properties": "warn",
    "no-new": "warn",
    "no-unneeded-ternary": "error",
    "no-useless-constructor": "error",
    "no-named-as-default-member": "warn",
    "no-absolute-path": "error",
    "no-duplicates": "warn",
    "no-empty-named-blocks": "error",
    "no-self-import": "error",
    "no-commented-out-tests": "warn",
    "approx-constant": "error",
    "misrefactored-assign-op": "warn",
    "no-promise-in-callback": "warn",
    "no-confusing-non-null-assertion": "error",
    "no-unnecessary-type-constraint": "warn",
    "no-extraneous-class": "error",
    "consistent-function-scoping": "error",
    "no-accessor-recursion": "error",
    "prefer-add-event-listener": "error",
    "array-callback-return": "warn",
    "eqeqeq": "warn",
    "max-classes-per-file": "warn",
    "no-array-constructor": "warn",
    "no-case-declarations": "error",
    "no-constructor-return": "error",
    "no-else-return": "error",
    "no-fallthrough": "warn",
    "no-lonely-if": "warn",
    "no-new-wrappers": "error",
    "no-object-constructor": "error",
    "no-prototype-builtins": "error",
    "no-self-compare": "error",
    "no-throw-literal": "error",
    "radix": "warn",
    "require-await": "warn",
    "symbol-description": "warn",
    "no-conditional-in-test": "warn",
    "require-param": "warn",
    "require-param-description": "warn",
    "require-param-name": "warn",
    "ban-ts-comment": "error",
    "no-unsafe-function-type": "warn",
    "prefer-ts-expect-error": "error",
    "escape-case": "warn",
    "no-hex-escape": "error",
    "no-instanceof-array": "error",
    "no-new-buffer": "error",
    "no-object-as-default-parameter": "error",
    "no-this-assignment": "error",
    "no-typeof-undefined": "warn",
    "no-unreadable-iife": "error",
    "no-useless-promise-resolve-reject": "warn",
    "no-useless-switch-case": "warn",
    "no-useless-undefined": "warn",
    "prefer-array-flat": "warn",
    "prefer-array-some": "warn",
    "prefer-blob-reading-methods": "error",
    "prefer-code-point": "warn",
    "prefer-date-now": "error",
    "prefer-dom-node-append": "warn",
    "prefer-dom-node-dataset": "warn",
    "prefer-dom-node-remove": "warn",
    "prefer-event-target": "warn",
    "prefer-math-trunc": "warn",
    "prefer-native-coercion-functions": "error",
    "prefer-prototype-methods": "error",
    "prefer-regexp-test": "warn",
    "prefer-string-replace-all": "warn",
    "prefer-string-slice": "error",
    "prefer-type-error": "warn",
    "require-number-to-fixed-digits-argument": "warn",
    "curly": "warn",
    "default-case-last": "error",
    "default-param-last": "error",
    "func-names": ["warn", "always"],
    "func-style": ["warn", "declaration"],
    "grouped-accessor-pairs": "error",
    "guard-for-in": "warn",
    "new-cap": "warn",
    "no-duplicate-imports": "error",
    "no-labels": "error",
    "no-magic-numbers": "warn",
    "no-return-assign": "warn",
    "no-multi-str": "warn",
    "no-nested-ternary": "error",
    "no-script-url": "error",
    "no-new-func": "error",
    "no-template-curly-in-string": "warn",
    "prefer-numeric-literals": "error",
    "prefer-object-spread": "warn",
    "prefer-promise-reject-errors": "error",
    "prefer-rest-params": "error",
    "prefer-spread": "error",
    "yoda": "error",
    "import/first": "error",
    "consistent-test-it": ["error", {"fn": "it", "withinDescribe": "it"}],
    "no-alias-methods": "error",
    "no-confusing-set-timeout": "error",
    "no-deprecated-functions": "warn",
    "no-done-callback": "warn",
    "no-duplicate-hooks": "warn",
    "no-identical-title": "warn",
    "no-restricted-jest-methods": "warn",
    "no-test-return-statement": "warn",
    "no-untyped-mock-factory": "warn",
    "prefer-comparison-matcher": "warn",
    "prefer-each": "warn",
    "prefer-equality-matcher": "error",
    "prefer-expect-resolves": "warn",
    "prefer-hooks-in-order": "warn",
    "prefer-hooks-on-top": "warn",
    "prefer-mock-promise-shorthand": "warn",
    "prefer-spy-on": "error",
    "prefer-to-be": "warn",
    "prefer-to-contain": "warn",
    "prefer-to-have-length": "warn",
    "prefer-todo": "warn",
    "require-hook": "error",
    "require-top-level-describe": "error",
    "avoid-new": "warn",
    "no-nesting": "warn",
    "no-return-wrap": "error",
    "param-names": "error",
    "prefer-await-to-callbacks": "warn",
    "prefer-catch": "error",
    "adjacent-overload-signatures": "warn",
    "array-type": "error",
    "consistent-generic-constructors": "error",
    "consistent-indexed-object-style": "error",
    "no-empty-interface": "warn",
    "no-inferrable-types": "warn",
    "prefer-for-of": "error",
    "prefer-function-type": "warn",
    "catch-error-name": "warn",
    "consistent-date-clone": "error",
    "consistent-existence-index-check": "error",
    "empty-brace-spaces": "error",
    "error-message": "error",
    "filename-case": "error",
    "no-console-spaces": "warn",
    "no-null": "error",
    "no-unreadable-array-destructuring": "error",
    "no-zero-fractions": "error",
    "numeric-separators-style": "error",
    "prefer-array-flat-map": "error",
    "prefer-includes": "error",
    "prefer-logical-operator-over-ternary": "error",
    "prefer-modern-dom-apis": "error",
    "prefer-structured-clone": "error",
    "switch-case-braces": "error",
    "throw-new-error": "error",
    "no-import-node-test": "error"
  }
}
```

</p>
</details> 

### What happened?

Given the following code :
```ts
// imports left out for brevity

function defineApplicationHook(name: keyof ApplicationHooks): (callback: ((app: Application) => void)) => void {
	return callback => {
		const app = getActiveApp();
		if (!app) {
			if (import.meta.env.DEV) {
				console.error(new Error(`${name} hook called without an active plugin instance`));
			}
			return;
		}

		app.emitter.on(name, callback);
	}
}
```

The arrow function is flagged as "Missing return type on function." even though the return type is provided by the `defineApplicationHook` function's return type (the typescript language server correctly infers it).

Doing the following changes clears the error, but it should not be necessary.
```diff
function defineApplicationHook(name: keyof ApplicationHooks): (callback: ((app: Application) => void)) => void {
- 	return callback => {
+ 	return (callback): void => {
		const app = getActiveApp();
```

### What should happen ?
The function should not be marked as "Missing return type on function." by the linter.
